File: 0424-longest-repeating-character-replacement/0424-longest-repeating-character-replacement.py
"""
Time = O(n)
Space = O(1)
"""
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        max_interval = 0
        counter = {}
        left = 0
        
        for right in range(len(s)):
            if s[right] not in counter:
                counter[s[right]] = 1
            else:
                counter[s[right]] += 1
            
            most_freq = max(counter.values())
            
            while (right - left + 1) - most_freq > k:
                counter[s[left]] -= 1
                if counter[s[left]] == 0:
                    counter.pop(s[left])
                left += 1
            
            max_interval = max(max_interval, right- left + 1)
            
        return max_interval 

File: 0424-longest-repeating-character-replacement/0424-longest-repeating-character-replacement.cpp
#include <vector>

class Solution {
public:
    int characterReplacement(string s, int k) {
        int left = 0, res = 0;
        // current highest frequency
        int curFreq;
        vector<int> counter(26, 0);
        for (int right; right < s.size(); right++) {
            // update counter while right pointer keep going right
            counter[s[right] - 'A']++;
            // find the highest frequency
            curFreq = *max_element(counter.begin(), counter.end());
            // move left pointer if it is not a valid substring for replacement
            while ((right - left + 1) - curFreq > k) {
                    counter[s[left] - 'A']--;
                    left++;
                    // update current highest frequency
                    curFreq = *max_element(counter.begin(), counter.end());
            }
            // update res
            res = max(res, right - left + 1);
        }
        return res;
    }
};